src/Eloquent/Model.php

```diff
@@ -77,698 +29,15 @@ abstract class Model extends BaseModel
-                $castValue = $this->serializeDate($castValue);
-            }
-
-            if ($castValue !== null && ($this->isCustomDateTimeCast($castType) || $this->isImmutableCustomDateTimeCast($castType))) {
-                $castValue = $castValue->format(explode(':', $castType, 2)[1]);
-            }
-
-            if ($castValue instanceof DateTimeInterface && $this->isClassCastable($key)) {
-                $castValue = $this->serializeDate($castValue);
-            }
-
-            if ($castValue !== null && $this->isClassSerializable($key)) {
-                $castValue = $this->serializeClassCastableAttribute($key, $castValue);
-            }
-
-            if ($this->isEnumCastable($key) && (! $castValue instanceof Arrayable)) {
-                $castValue = $castValue !== null ? $this->getStorableEnumValueFromLaravel11($this->getCasts()[$key], $castValue) : null;
-            }
-
-            if ($castValue instanceof Arrayable) {
-                $castValue = $castValue->toArray();
-            }
-
-            Arr::set($attributes, $key, $castValue);
-        }
-
-        return $attributes;
-    }
-
-    /**
-     * Duplicate of {@see HasAttributes::getStorableEnumValue()} for Laravel 11 as the signature of the method has
-     * changed in a non-backward compatible way.
-     *
-     * @todo Remove this method when support for Laravel 10 is dropped.
-     */
-    private function getStorableEnumValueFromLaravel11($expectedEnum, $value)
-    {
-        if (! $value instanceof $expectedEnum) {
-            throw new ValueError(sprintf('Value [%s] is not of the expected enum type [%s].', var_export($value, true), $expectedEnum));
-        }
-
-        return $value instanceof BackedEnum
-            ? $value->value
-            : $value->name;
-    }
-
-    /**
-     * Is a value a BSON type?
-     *
-     * @param mixed $value
-     *
-     * @return bool
-     */
-    protected function isBSON(mixed $value): bool
-    {
-        return $value instanceof Type;
-    }
-
-    /**
-     * {@inheritDoc}
-     */
-    public function save(array $options = [])
-    {
-        // SQL databases would use autoincrement the id field if set to null.
-        // Apply the same behavior to MongoDB with _id only, otherwise null would be stored.
-        if (array_key_exists('_id', $this->attributes) && $this->attributes['_id'] === null) {
-            unset($this->attributes['_id']);
-        }
-
-        $saved = parent::save($options);
-
-        // Clear list of unset fields
-        $this->unset = [];
-
-        return $saved;
-    }
-
-    /**
-     * {@inheritDoc}
-     */
-    public function refresh()
-    {
-        parent::refresh();
-
-        // Clear list of unset fields
-        $this->unset = [];
-
-        return $this;
+        return is_subclass_of($related, BaseModel::class)
+            && array_key_exists(DocumentModel::class, class_uses_recursive($related));
```

Comment:
`class_uses_recursive` looks like it has significant performance implications. Using a marker interface and checking using `is_subclass_of` would probably be more efficient.

Alternatively, checking whether the class is a subclass of this model class first, and only checking for traits in other cases would probably solve the performance issue for most cases.

Reply:
Added a shortcut to check our Model class + a memory cache for when the same class is checked twice in the same request.